src/hooks/userspace/hook-grpc.js

```diff
@@ -224,46 +204,48 @@ function wrapUnary(namespace, handlerSet, requestName) {
  * @param {string} requestName The human-friendly name of the request.
  */
 function wrapServerStream(namespace, handlerSet, requestName) {
-  var rootContext;
-  // CLS context object.
-  var context = {};
-  // Start right before deserialization (which precedes call to function)
-  // End right after output stream 'finish' event
-  shimmer.wrap(handlerSet, 'deserialize', function (deserialize) {
-    return namespace.bind(function deserializeTrace() {
-      rootContext = startRootSpanForRequest(requestName);
-      return deserialize.apply(this, arguments);
-    }, context);
-  });
   shimmer.wrap(handlerSet, 'func', function (func) {
-    return namespace.bind(function funcTrace(call) {
-      if (agent.config_.enhancedDatabaseReporting) {
-        shimmer.wrap(call, 'sendMetadata', sendMetadataWrapper(rootContext));
-      }
-      if (agent.config_.enhancedDatabaseReporting) {
-        rootContext.addLabel('argument', JSON.stringify(call.request));
-      }
-      var spanEnded = false;
-      var endSpan = function () {
-        if (!spanEnded) {
-          spanEnded = true;
-          endRootSpanForRequest(rootContext);
-        }
-      };
-      call.on('finish', function () {
-        // End the span unless there is an error.
-        if (call.status.code === 0) {
-          endSpan();
+    return function serverMethodTrace(stream) {
+      var that = this;
+      var args = arguments;
+      // Running in the namespace here propagates context to func.
+      return namespace.runAndReturn(function() {
+        var rootContext = startRootSpanForRequest(requestName, 5);
+        if (agent.config_.enhancedDatabaseReporting) {
+          shimmer.wrap(stream, 'sendMetadata', sendMetadataWrapper(rootContext));
         }
-      });
-      call.on('error', function (err) {
         if (agent.config_.enhancedDatabaseReporting) {
-          rootContext.addLabel('error', err);
+          rootContext.addLabel('argument', JSON.stringify(stream.request));
         }
-        endSpan();
+        var spanEnded = false;
+        var endSpan = function () {
+          if (!spanEnded) {
+            spanEnded = true;
+            endRootSpanForRequest(rootContext);
+          }
+        };
+        // Propagate context to stream event handlers.
+        namespace.bindEmitter(stream);
+        // stream is a WriteableStream. Emitting a 'finish' or 'error' event
+        // suggests that no more data will be sent, so we end the span in these
+        // event handlers.
+        stream.on('finish', function () {
+          // End the span unless there is an error. (If there is, the span will
+          // be ended in the error event handler. This is to ensure that the
+          // 'error' label is applied.)
+          if (stream.status.code === 0) {
+            endSpan();
+          }
+        });
+        stream.on('error', function (err) {
```

Comment:
`on` actually returns the instance for chaining so one could do:

`eventEmitter.on(...).on(...);`